src/box/gc.h

```diff
@@ -223,6 +223,13 @@ gc_last_checkpoint(void)
 void
 gc_init(on_garbage_collection_f on_garbage_collection);
 
+/**
+ * Prepare for freeing resources in gc_free while TX event loop is
+ * still running.
+ */
+void
+gc_shutdown(void);
+
```

Comment:
What makes sure a new checkpoint isn't started between `shutdown` and `free`?

Reply:
We do `gc_shutdown` after all client fibers are finished. Added assertions to gc API as discussed f2f. Only to the functions where things can go wrong (a bit asymetrical though).

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/gc.c b/src/box/gc.c
index 41da69179..d84aee68c 100644
--- a/src/box/gc.c
+++ b/src/box/gc.c
@@ -322,6 +322,7 @@ gc_cleanup_fiber_f(va_list ap)
 void
 gc_set_wal_cleanup_delay(double wal_cleanup_delay)
 {
+	assert(gc.cleanup_fiber != NULL);
 	gc.wal_cleanup_delay = wal_cleanup_delay;
 	/*
 	 * This routine may be called at arbitrary
@@ -345,6 +346,7 @@ gc_delay_ref(void)
 void
 gc_delay_unref(void)
 {
+	assert(gc.cleanup_fiber != NULL);
 	if (gc.is_paused) {
 		assert(gc.delay_ref > 0);
 		gc.delay_ref--;
@@ -388,6 +390,7 @@ gc_wait_cleanup(void)
 void
 gc_advance(const struct vclock *vclock)
 {
+	assert(gc.cleanup_fiber != NULL);
 	/*
 	 * In case of emergency ENOSPC, the WAL thread may delete
 	 * WAL files needed to restore from backup checkpoints,
@@ -431,6 +434,7 @@ gc_set_min_checkpoint_count(int min_checkpoint_count)
 void
 gc_set_checkpoint_interval(double interval)
 {
+	assert(gc.checkpoint_fiber != NULL);
 	/*
 	 * Reconfigure the schedule and wake up the checkpoint
 	 * daemon so that it can readjust.
@@ -448,6 +452,7 @@ gc_set_checkpoint_interval(double interval)
 void
 gc_add_checkpoint(const struct vclock *vclock)
 {
+	assert(gc.cleanup_fiber != NULL);
 	struct gc_checkpoint *last_checkpoint = gc_last_checkpoint();
 	if (last_checkpoint != NULL &&
 	    vclock_sum(&last_checkpoint->vclock) == vclock_sum(vclock)) {
@@ -538,6 +543,7 @@ gc_do_checkpoint(bool is_scheduled)
 int
 gc_checkpoint(void)
 {
+	assert(gc.checkpoint_fiber != NULL);
 	if (gc.checkpoint_is_in_progress) {
 		diag_set(ClientError, ER_CHECKPOINT_IN_PROGRESS);
 		return -1;
@@ -577,6 +583,7 @@ gc_checkpoint(void)
 void
 gc_trigger_checkpoint(void)
 {
+	assert(gc.checkpoint_fiber != NULL);
 	if (gc.checkpoint_is_in_progress || gc.checkpoint_is_pending)
 		return;
 
@@ -648,6 +655,7 @@ gc_ref_checkpoint(struct gc_checkpoint *checkpoint,
 void
 gc_unref_checkpoint(struct gc_checkpoint_ref *ref)
 {
+	assert(gc.cleanup_fiber != NULL);
 	rlist_del_entry(ref, in_refs);
 	gc_schedule_cleanup();
 }
@@ -675,6 +683,7 @@ gc_consumer_register(const struct vclock *vclock, const char *format, ...)
 void
 gc_consumer_unregister(struct gc_consumer *consumer)
 {
+	assert(gc.cleanup_fiber != NULL);
 	if (!consumer->is_inactive) {
 		gc_tree_remove(&gc.consumers, consumer);
 		gc_schedule_cleanup();
@@ -685,6 +694,7 @@ gc_consumer_unregister(struct gc_consumer *consumer)
 void
 gc_consumer_advance(struct gc_consumer *consumer, const struct vclock *vclock)
 {
+	assert(gc.cleanup_fiber != NULL);
 	if (consumer->is_inactive)
 		return;
 

```

</details>

Reply:
Also zero-out fiber variables right after join as suggested f2f.

```diff
diff --git a/src/box/gc.c b/src/box/gc.c
index d84aee68c..ed4f381a4 100644
--- a/src/box/gc.c
+++ b/src/box/gc.c
@@ -141,8 +141,8 @@ gc_shutdown(void)
        fiber_cancel(gc.checkpoint_fiber);
        fiber_cancel(gc.cleanup_fiber);
        fiber_join(gc.checkpoint_fiber);
-       fiber_join(gc.cleanup_fiber);
        gc.checkpoint_fiber = NULL;
+       fiber_join(gc.cleanup_fiber);
        gc.cleanup_fiber = NULL;
 }
```